v0/destinations/mailchimp/transform.js

```diff
@@ -1,107 +1,153 @@
 } = require("../../util");
-const logger = require("../../../logger");
-
-// Converts to upper case and removes spaces
-function filterTagValue(tag) {
-  const maxLength = 10;
-  const newTag = tag.replace(/[^\w\s]/gi, "");
-  if (newTag.length > maxLength) {
-    return newTag.slice(0, 10);
-  }
-  return newTag.toUpperCase();
-}
+const {
+  getMailChimpEndpoint,
+  filterTagValue,
+  checkIfMailExists,
+  checkIfDoubleOptIn
+} = require("./utils");
 
-function getMailChimpEndpoint(mailChimpConfig) {
-  return getEndpoint(mailChimpConfig.dataCenterId, mailChimpConfig.audienceId);
-}
+const { MappedToDestinationKey } = require("../../../constants");
+const {
+  MAX_BATCH_SIZE,
+  MAILCHIMP_IDENTIFY_EXCLUSION,
+  SUBSCRIPTION_STATUS,
+  VALID_STATUSES,
+  mergeConfig
+} = require("./config");
 
-async function checkIfMailExists(mailChimpConfig, email) {
-  if (!email) {
-    return false;
-  }
-  const hash = md5(email);
-  const url = `${getMailChimpEndpoint(mailChimpConfig)}/members/${hash}`;
-
-  let status = false;
-  try {
-    await axios.get(url, {
-      auth: {
-        username: "apiKey",
-        password: `${mailChimpConfig.apiKey}`
+const mergeAdditionalTraitsFields = (traits, mergedFieldPayload) => {
+  if (isDefined(traits)) {
+    Object.keys(traits).forEach(trait => {
+      // if any trait field is present, other than the fixed mapping, that is passed as well.
+      if (MAILCHIMP_IDENTIFY_EXCLUSION.indexOf(trait) === -1) {
+        const tag = filterTagValue(trait);
+        mergedFieldPayload[tag] = traits[trait];
       }
     });
-    status = true;
-  } catch (error) {
-    logger.error("axios error");
   }
-  return status;
-}
+  return mergedFieldPayload;
+};
 
-async function checkIfDoubleOptIn(mailChimpConfig) {
-  let response;
-  const url = `${getMailChimpEndpoint(mailChimpConfig)}`;
-  try {
-    response = await axios.get(url, {
-      auth: {
-        username: "apiKey",
-        password: `${mailChimpConfig.apiKey}`
+const processPayloadBuild = async (
+  message,
+  updateSubscription,
+  primaryPayload,
+  Config,
+  emailExists,
+  audienceId,
+  enableMergeFields
+) => {
+  const traits = getFieldValueFromMessage(message, "traits");
+  const email = getFieldValueFromMessage(message, "email");
+  // ref: https://mailchimp.com/developer/marketing/docs/merge-fields/#structure
+  const mergedFieldPayload = constructPayload(message, mergeConfig);
+  const { apiKey, datacenterId } = Config;
+  let allMergedFields;
+
+  // for sending any fields other than email_address, while the email is already existing
+  // enableMergeFields needs to be set to true.
+  if (isDefinedAndNotNull(updateSubscription) && emailExists) {
+    if (isDefined(enableMergeFields) && enableMergeFields === true) {
+      allMergedFields = mergeAdditionalTraitsFields(traits, mergedFieldPayload);
+    } else {
+      allMergedFields = null;
+    }
+  } else {
+    // if user sends a non-existing email, the trait fields are sent within
```

Comment:
According to our previous implementation, user consent was only taken for merging fields, when the email already exists. But if it doesn’t it sent using the merge field options. Should we change that? @utsabc

Reply:
Got it I mean we should change that but it would be a breaking change then. And no way we can keep this backward compatible. @saikumarrs any thoughts here?